File: GameConfig/Lua/M_StarLotteryCFG.lua
M_StarLotteryCFG = {}

function M_StarLotteryCFG:getData(key)
	if self.datas == nil then
		return nil
	end
	return self.datas[key]
end

function M_StarLotteryCFG:init()
	self.datas = {}
	self.datas[1] = {Index = 1, StarLvl = 1, ItemType = 1, ItemCount = 200, Rate1 = 10000, Rate2 = 10000, Rate3 = 10000, Rate4 = 10000}
	self.datas[2] = {Index = 2, StarLvl = 1, ItemType = 1, ItemCount = 2000, Rate1 = 3850, Rate2 = 2775, Rate3 = 2517, Rate4 = 2280}
	self.datas[3] = {Index = 3, StarLvl = 1, ItemType = 2, ItemCount = 2, Rate1 = 2350, Rate2 = 1375, Rate3 = 1217, Rate4 = 1080}
	self.datas[4] = {Index = 4, StarLvl = 1, ItemType = 2, ItemCount = 5, Rate1 = 1550, Rate2 = 625, Rate3 = 517, Rate4 = 430}
	self.datas[5] = {Index = 5, StarLvl = 1, ItemType = 11, ItemCount = 1, Rate1 = 1000, Rate2 = 125, Rate3 = 67, Rate4 = 30}
	self.datas[6] = {Index = 6, StarLvl = 2, ItemType = 1, ItemCount = 1000, Rate1 = 10000, Rate2 = 10000, Rate3 = 10000, Rate4 = 10000}
	self.datas[7] = {Index = 7, StarLvl = 2, ItemType = 1, ItemCount = 10000, Rate1 = 2851, Rate2 = 2401, Rate3 = 2101, Rate4 = 1821}
	self.datas[8] = {Index = 8, StarLvl = 2, ItemType = 2, ItemCount = 10, Rate1 = 851, Rate2 = 501, Rate3 = 301, Rate4 = 121}
	self.datas[9] = {Index = 9, StarLvl = 2, ItemType = 2, ItemCount = 50, Rate1 = 451, Rate2 = 201, Rate3 = 101, Rate4 = 21}
	self.datas[10] = {Index = 10, StarLvl = 2, ItemType = 11, ItemCount = 3, Rate1 = 300, Rate2 = 100, Rate3 = 50, Rate4 = 20}
	self.datas[11] = {Index = 11, StarLvl = 3, ItemType = 1, ItemCount = 5000, Rate1 = 10000, Rate2 = 10000, Rate3 = 10000, Rate4 = 10000}
	self.datas[12] = {Index = 12, StarLvl = 3, ItemType = 1, ItemCount = 50000, Rate1 = 1550, Rate2 = 1280, Rate3 = 990, Rate4 = 710}
	self.datas[13] = {Index = 13, StarLvl = 3, ItemType = 2, ItemCount = 20, Rate1 = 750, Rate2 = 580, Rate3 = 390, Rate4 = 210}
	self.datas[14] = {Index = 14, StarLvl = 3, ItemType = 2, ItemCount = 150, Rate1 = 300, Rate2 = 230, Rate3 = 140, Rate4 = 60}
	self.datas[15] = {Index = 15, StarLvl = 3, ItemType = 11, ItemCount = 10, Rate1 = 100, Rate2 = 80, Rate3 = 40, Rate4 = 10}
	self.datas[16] = {Index = 16, StarLvl = 4, ItemType = 1, ItemCount = 10000, Rate1 = 10000, Rate2 = 10000, Rate3 = 10000, Rate4 = 10000}
	self.datas[17] = {Index = 17, StarLvl = 4, ItemType = 1, ItemCount = 100000, Rate1 = 2870, Rate2 = 3265, Rate3 = 3100, Rate4 = 2955}
	self.datas[18] = {Index = 18, StarLvl = 4, ItemType = 2, ItemCount = 50, Rate1 = 870, Rate2 = 765, Rate3 = 600, Rate4 = 455}
	self.datas[19] = {Index = 19, StarLvl = 4, ItemType = 2, ItemCount = 300, Rate1 = 370, Rate2 = 340, Rate3 = 250, Rate4 = 180}
	self.datas[20] = {Index = 20, StarLvl = 4, ItemType = 11, ItemCount = 20, Rate1 = 50, Rate2 = 70, Rate3 = 30, Rate4 = 10}
	self.datas[21] = {Index = 21, StarLvl = 5, ItemType = 1, ItemCount = 80000, Rate1 = 10000, Rate2 = 10000, Rate3 = 10000, Rate4 = 10000}
	self.datas[22] = {Index = 22, StarLvl = 5, ItemType = 1, ItemCount = 500000, Rate1 = 2125, Rate2 = 2420, Rate3 = 2220, Rate4 = 2020}
	self.datas[23] = {Index = 23, StarLvl = 5, ItemType = 2, ItemCount = 100, Rate1 = 1125, Rate2 = 920, Rate3 = 720, Rate4 = 520}
	self.datas[24] = {Index = 24, StarLvl = 5, ItemType = 2, ItemCount = 500, Rate1 = 525, Rate2 = 420, Rate3 = 320, Rate4 = 220}
	self.datas[25] = {Index = 25, StarLvl = 5, ItemType = 11, ItemCount = 50, Rate1 = 25, Rate2 = 20, Rate3 = 20, Rate4 = 20}
end

M_StarLotteryCFG:init()

File: GameConfig/Code/Client/M_StarLotteryCFG.cpp
#include "stdafx.h"
#include "tinyxml2/tinyxml2.h"
#include "LuaCfgHelper.h"
#include <iostream>
#include <boost/algorithm/string.hpp>
#include "M_StarLotteryCFG.h"
std::auto_ptr<M_StarLotteryCFG> M_StarLotteryCFG::msSingleton(nullptr);

int M_StarLotteryCFG::GetCount()
{
	return (int)mMapData.size();
}

const M_StarLotteryCFGData* M_StarLotteryCFG::GetData(int Index)
{
	auto it = mMapData.find(Index);
	if (it != mMapData.end())
	{
		return &it->second;
	}
	return NULL;
}

const std::map<int, M_StarLotteryCFGData>& M_StarLotteryCFG::GetMapData()
{
	return mMapData;
}

void M_StarLotteryCFG::Load()
{
	tinyxml2::XMLDocument xmlDoc;
	std::string content = FileUtils::getInstance()->getStringFromFile("Config/M_StarLotteryCFG.xml");
	auto result = xmlDoc.Parse(content.c_str(), content.length());
	if (result != tinyxml2::XML_SUCCESS)
	{
		CCLOGERROR("Result:%d", result);
		CCASSERT(false, "result != tinyxml2::XML_SUCCESS");
		return;
	}
	auto root = xmlDoc.RootElement();
	if (root == NULL)
	{
		CCASSERT(false, "root == NULL");
		return;
	}
	auto element = root->FirstChildElement("Data");
	while (element != NULL)
	{
		M_StarLotteryCFGData data;
		data.mIndex = element->IntAttribute("Index");
		data.mStarLvl = element->IntAttribute("StarLvl");
		data.mItemType = element->IntAttribute("ItemType");
		data.mItemCount = element->IntAttribute("ItemCount");
		data.mRate1 = element->IntAttribute("Rate1");
		data.mRate2 = element->IntAttribute("Rate2");
		data.mRate3 = element->IntAttribute("Rate3");
		data.mRate4 = element->IntAttribute("Rate4");
		if (mMapData.find(data.mIndex) != mMapData.end())std::cout <<"data refind:" << data.mIndex << std::endl;
		CCASSERT(mMapData.find(data.mIndex) == mMapData.end(), "data.mIndex is exists");
		mMapData.insert(std::make_pair(data.mIndex, data));
		element = element->NextSiblingElement();
	}
	CCLOG("M_StarLotteryCFG Loaded. Load Data:%u", mMapData.size());
}

void M_StarLotteryCFG::LoadLua()
{
	LuaEngine::getInstance()->executeScriptFile("config/M_StarLotteryCFG");
	lua_State* L = LuaEngine::getInstance()->getLuaStack()->getLuaState();
	lua_getglobal(L, "M_StarLotteryCFG");
	CCASSERT(lua_istable(L, -1) == 1, "is not table");
	lua_pushstring(L, "datas");
	lua_gettable(L, -2);
	CCASSERT(lua_istable(L, -1) == 1, "is not table");
	lua_pushnil(L);
	while(lua_next(L, 2))
	{
		CCASSERT(lua_istable(L, -1) == 1, "is not table");
		M_StarLotteryCFGData data;
		LuaCfgHelper::readInt(L, "Index", data.mIndex);
		LuaCfgHelper::readInt(L, "StarLvl", data.mStarLvl);
		LuaCfgHelper::readInt(L, "ItemType", data.mItemType);
		LuaCfgHelper::readInt(L, "ItemCount", data.mItemCount);
		LuaCfgHelper::readInt(L, "Rate1", data.mRate1);
		LuaCfgHelper::readInt(L, "Rate2", data.mRate2);
		LuaCfgHelper::readInt(L, "Rate3", data.mRate3);
		LuaCfgHelper::readInt(L, "Rate4", data.mRate4);
		if (mMapData.find(data.mIndex) != mMapData.end())std::cout <<"data refind:" << data.mIndex << std::endl;
		CCASSERT(mMapData.find(data.mIndex) == mMapData.end(), "data.mIndex is exists");
		mMapData.insert(std::make_pair(data.mIndex, data));
		lua_pop(L, 1);
	}
	lua_settop(L, 0);
	CCLOG("M_StarLotteryCFG Loaded. Load Data:%u", mMapData.size());
}

void M_StarLotteryCFG::Reload()
{
	mMapData.clear();
	Load();
}

M_StarLotteryCFG* M_StarLotteryCFG::GetSingleton()
{
	if (msSingleton.get() == nullptr)
	{
		msSingleton.reset(new M_StarLotteryCFG());
	}
	return msSingleton.get();
}

void M_StarLotteryCFG::Release()
{
	msSingleton.reset(nullptr);
}
